examples-gallery/plot_ball_rolling_on_spinning_disc.py

```diff
@@ -0,0 +1,620 @@
+    )
+
+(fr, frstar) = KM.kanes_equations(bodies, forces)
+eom = kd.col_join(fr + frstar)
+eom = me.msubs((eom.col_join(hol_constr)),
+    {sm.Derivative(T(t), t):
+    Tdot, sm.Derivative(T(t), (t,2)): Tdotdot},
+    )
+print(f'eom contains {sm.count_ops(eom):,} operations, ' +
+    f'{sm.count_ops(sm.cse(eom))} after cse')
+
+# %%
+# Set up the optimization problem and solve it.
+
+h = sm.symbols('h')
+state_symbols = (*q_ind, *q_dep, *u_ind, *u_dep)
+laenge = len(state_symbols)
+constant_symbols = (r, mb, mo, g, Omega, alpha, Tdot, Tdotdot)
+specified_symbols = (t1, t2, t3)
+methode = "backward euler"
+num_nodes = 250
+duration = (num_nodes - 1) * h
+
+# %%
+# Disc time is a constant, related to the acceleration of the disc.
+disc_time = 7.5
+
+# %%
+interval_value = h
+interval_value_fix = disc_time/num_nodes
+interval_fix = np.linspace(0, disc_time, num_nodes)
+
+# %%
+# Specify the known system parameters.
+par_map = {}
+par_map[mb]  = 5.0
+par_map[mo]  = 1.0
+par_map[r]   = 1.0
+par_map[Omega] = 10.0
+par_map[alpha] = 0.5
+par_map[g]   = 9.81
+par_map[Tdot] = interval_value_fix
+par_map[Tdotdot] = 0.0
+
+# %%
+# A weighted sum of energy and speed is to be minimized. weight is the
+# relative weight of the speed. (If weight = 0, only the energy is
+# minimized.)
+weight = 2.5e5
+
+# %%
+# Define the objective function and its gradient.
+def obj(free):
+    free1 = free[0: -1]
+    Tz1 = free1[laenge * num_nodes: (laenge + 1) * num_nodes]
+    Tz2 = free1[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes]
+    Tz3 = free1[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes]
+    return free[-1] * (np.sum(Tz1**2) + np.sum(Tz2**2) +
+        np.sum(Tz3**2)) + free[-1] * weight
+
+def obj_grad(free):
+    grad = np.zeros_like(free)
+    grad[laenge * num_nodes: (laenge + 1) * num_nodes] = (2.0 * free[-1]
+         * free[laenge * num_nodes: (laenge + 1) * num_nodes])
+    grad[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes] = (2.0 *
+        free[-1] * free[(laenge + 1) * num_nodes: (laenge + 2) * num_nodes])
+    grad[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes] = (2.0 * free[-1]
+       * free[(laenge + 2) * num_nodes: (laenge + 3) * num_nodes])
+    grad[-1] = weight
+    return grad
+
+t0, tf = 0.0, duration
+
+# %%
+# The holonomic constraints must be fullfilled by the state constraints.
+x_start = 7.0
+q2_start = x_start / par_map[r]
+y_start = 7.0
+q1_start = -y_start / par_map[r]
+
+initial_state_constraints = {
+    q1: q1_start,
+    q2: q2_start,
+    q3: 0.0,
+    u1: 0.0,
+    u2: 0.0,
+    u3: 0.0,
+    x: x_start,
+    y: y_start,
+    ux: 0.0,
+    uy: 0.0
+    }
+final_state_constraints = {
+    x: 0.0,
+    y: 0.0,
+    ux: 0.0,
+    uy: 0.0,
+    }
+
+instance_constraints = (
+) + tuple(
+    xi.subs({t: t0}) - xi_val for xi, xi_val in
+    initial_state_constraints.items()
+) + tuple(
+    xi.subs({t: tf}) - xi_val for xi, xi_val in
+    final_state_constraints.items()
+)
```

Comment:
I changed this in your other PR, it seems unnecessarily complex.

Reply:
Also syntatic sugar. Explicit is `instance_constraints = (x(1.0) - 0.0, ...)`

Reply:
> Also syntatic sugar. Explicit is `instance_constraints = (x(1.0) - 0.0, ...)`

I'd copied this from Timo. Once I understood it I loved it - but I understand what you mean. Will be change like all the otehr suggestions.